<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Привет! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Мои комментарии обозначены пометкой **Комментарий наставника**. Далее в файле ты сможешь найти их в похожих ячейках (если рамки комментария зелёные - всё сделано правильно; оранжевые - есть замечания, но не критично; красные - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [2]:
import pandas as pd    # Импорт библиотеки pandas
data = pd.read_csv('/datasets/data.csv')    # Чтение файла с данными функцией read_csv и запись данных в переменную data


data.info()    # Просмотр общей информации о таблице данных методом info(), определение количества и типов данных в таблице
data.head(10)    # Просмотр первых десяти значений в таблице методом head с аргументом 10, оценка качества значений

# Далее приводятся примененные фукнции и методы для более детального изучения данных. Исполняются последовательно путём снятия/установки хэштэга, т.е. знака комментария
# Прямого отношения к решению поставленных в проекте задач данные функции не имеют, поэтому и закомментированы. Удалять жалко.
#print(data['family_status'].unique())    # Выявление уникальных значений в столбце "семейное положение". Для других столбцов с категорийными данными тоже следует использовать
#print(data['gender'].unique())    # Выявление уникальных значений в столбце "пол"
#print(data['income_type'].unique())    # Выявление уникальных значений в столбце "тип дохода"
#print(data['purpose'].unique())    # Выявление уникальных значений в столбце "цель получения кредита"
#print(data[data['gender'] == 'XNA'].head(10))    # Поиск личностей неопределившихся со своим полом
#print(data[(data['days_employed'].isna() == True) & (data['total_income'].isna() == True)])    # Проверка предположения о том, что для строк в которых значение стобца days_employed является пустым соответствующие значения стобца total_income также является пустым
#print(data['children'].max())    # Определение максимального значения в столбце "количество детей"
#print(data['days_employed'].max())   # Определение максимального значения в столбце "Общий трудовой стаж в днях"
#print(len(data['purpose'].unique()))    # Выявление количества уникальных значений в столбце "цель получения кредита"
#print(data['family_status'].value_counts())    # Подсчет количества клиентов с разными семейными положениями

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

### Вывод

### После ознакомления с исходными данными можно сделать следующие выводы:

1. Данные существуют и с ними можно работать! (файл data.csv не битый, спасибо заказчику)
1. В данных **21525** строк. Это примерно население города Хотьково Московской области по данным переписи на 2019 г. На мой взгляд данных для статистики мало. На этом можно было бы и закончить исследование, объяснив заказчику, что 21 тысяча частных клиентов это несерьёзно, и что его банк скорее всего не переживёт последствий экономического кризиса, следовательно, оплату за данное исследование мы не получим. Но принимая во внимание, что задача носит учебный характер, в дальнейшем буду стараться делать выводы только в рамках учебного процесса.
1. Столбец **children** содержит 21525 ненулевых значений типа ***int64*** - идеально. Принимает ***целочисленные*** значения от 0 до 20. Пропусков в столбце нет. Следует проверить на выбросы в виде отрицательных значений, а также чрезмерно больших значений, которые также будут нарушать статистику. Возможно при дальнейшем анализе имеет смысл применить к столбцу категоризацию следующего вида: 0 - нет детей, от 1 до 3 - есть дети, от 4 и более - многодетная семья.
1. Столбец **days_employed** содержит 19351 ненулевых значений типа ***float64*** - неидеально. Принимает ***вещественные***  значения. Есть пропущенные значения в 2174 строках, их нужно будет заполнить.‬ Не понятно зачем **общий трудовой стаж клиентов в днях** хранить в ***вещественном*** типе данных. Следует преобразовать в ***int64***. Также по значениям столбца видны отрицательные значения. Понаберут кредитов вот и работают в минус! На самом деле тут вероятно произошёл технический сбой при выгрузке данных, поэтому значения стали отрицательными. Следует как следует обработать данный столбец, но не слишком долго, ведь для решения поставленных проектом задач данный столбец не интересен.
1. Столбец **dob_years** содержит 21525 ненулевых значений типа ***int64*** - идеально. Принимает ***целочисленные*** значения. Для решения поставленных проектом задач данный столбец не интересен.
1. Столбец **education** содержит 21525 ненулевых значения типа  ***object*** - идеально. Принимает ***строковые*** значения об уровне образования клиентов. Одни и те же по смыслу значения записаны по-разному, а именно - не соблюдается однотипное форматирование (регистр). Следует привести все одинаковые по смыслу значения к одному виду, чтобы была возможность сосчитать количество клиентов с разным уровнем образования и выявить возможные дубликаты. Для решения поставленных проектом задач данный столбец не интересен.
1. Столбец **education_id** содержит 21525 ненулевых значений типа ***int64*** - идеально. Принимает ***целочисленные*** значения от 0 до 4. Путем сопоставления значений столбцов **education_id** и **education** выявлены следующие соответствия: 0 - Высшее, 1 - Среднее, 2 - Неоконченное высшее, 3 - Начальное, 4 - Ученая степень. Для решения поставленных проектом задач данный столбец не интересен.
1. Столбец **family_status** содержит 21525 ненулевых значения типа ***object*** - идеально. Принимает ***строковые*** значения: 'женат / замужем', 'гражданский брак', 'вдовец / вдова', 'в разводе', 'Не женат / не замужем'. Применение метода unique() к данному столбцу подтверждает, что значения в нём сразу же пригодны для дальнейшего анализа, т.к. не содержат одинаковых по смыслу, но разных по содержанию или формату значений, как например, в столбце ***education***. Значение 'Не женат / не замужем' почему-то записали с заглавной буквы в отличии от остальных значений. Пожалуй оставлю как есть - в глаза не бросается. Данный столбец важен для решения поставленных проектом задач.
1. Столбец **family_status_id**  содержит 21525 ненулевых значений типа ***int64*** - идеально. Принимает ***целочисленные*** значения от 0 до 4. Путем сопоставления значений столбцов **family_status_id** и **family_status** выявлены следующие соответствия: 0 - женат / замужем, 1 - гражданский брак, 2 - вдовец / вдова, 3 - в разводе, 4 - Не женат / не замужем. Данный столбец важен для решения поставленных проектом задач.
1. Столбец **gender** содержит 21525 ненулевых значения типа ***object*** - идеально. Принимает ***строковые*** значения: 'F' 'M' 'XNA'. Последнее значение **'XNA'** встречается в данных только один раз. Явно уникальная личность. Дальнейшие рассуждения на данную тему бессмысленны, поскольку для решения поставленных проектом задач данный столбец не интересен. Жаль.
1. Столбец **income_type** содержит 21525 ненулевых значения типа ***object*** - идеально. Принимает ***строковые*** значения: 'сотрудник', 'пенсионер', 'компаньон', 'госслужащий', 'безработный', 'предприниматель', 'студент', 'в декрете'. Значения выглядят вполне опрятно. Никаких действий применительно к данному столбцу выполнять не следует. Для решения поставленных проектом задач данный столбец не интересен.
1. Столбец **debt** содержит 21525 ненулевых значений типа ***int64*** - идеально. Принимает ***целочисленные*** значения 0 и 1. Путём логического рассуждения о назначении данного столбца, а именно - имел ли клиент задолженность по возврату кредитов, следует, что 0 - не имел, 1 - имел. Снова встает вопрос о нерациональном расходовании вычислительных ресурсов. Для столбцов которые могут принимать только два значения следовало бы использовать логический тип данный **bool**. Данный столбец важен для решения поставленных проектом задач.
1. Столбец **total_income** содержит 19351 ненулевых значений типа ***float64*** - неидеально. Принимает ***вещественные*** положительные значения. Есть пропущенные значения в 2174 строках, их нужно будет заполнить. Причём пропущенные значения в данном столбце встречаются в тех же строках, что и пропущенные значения в столбце **days_employed**. Данный столбец важен для решения поставленных проектом задач.
1. Столбец **purpose** содержит 21525 ненулевых значения типа ***object*** - идеально. Принимает ***строковые*** значения в количестве 38 уникальных значений. Это много. Учитывая, что многие значения по смыслу соответствуют одной и той же цели, следует выполнить категоризацию, чтобы уменьшить количество уникальных значений. Для этого можно применить лемматизацию, чтобы скорее отыскать повторяющиеся слова и по ним составить категории. Данный столбец важен для решения поставленных проектом задач.


<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

У меня имеется несколько комментариев по данному шагу:
- считывание данных и импорт необходимых библиотек лучше проводить в разных ячейках;
- следует давать переменным осознанные имена. Таблица про клиентов - название надо бы выбрать соответствующее;
- таблицы на печать выводи с помощью метода display;
- Подумай также о возможных причинах появления пропусков, а также о том, являются ли они случайными или нет.

Первый взгляд на таблицу выполнен. Здорово, что обнаружены ошибки в данных.

### Шаг 2. Предобработка данных

### Обработка пропусков

1. Определить и заполнить пропущенные значения;
1. Описать, какие пропущенные значения обнаружены;
1. Привести возможные причины появления пропусков в данных;
1. Объяснить, по какому принципу заполнены пропуски.

In [4]:
# Сначала следует понять насколько много пропущенных значений по сравнению с размером таблицы данных
data_rows_total = data.shape[0]    # Определение общего количества строк в исходной таблице и запись в переменную data_rows_total
print('Общее количество строк в исходной таблице составляет:', data_rows_total)   # Вывод на экран переменной data_rows_total

data_rows_na = data['total_income'].isna().sum()    # Определение количества пропущенных значений в столбце total_income и запись в переменную
print('Общее количество пропущенных значений в столбце "ежемесячный доход" составляет:', data_rows_na)   # Вывод на экран переменной data_rows_total

try:    # Использование конструкции try-except чтобы защититься от деления на ноль
    data_percent_na = data_rows_na / data_rows_total    # Вычисление отношения количества пропущенных значений к общему количеству строк таблицы
    print('Число пропущенных значений в исходной таблице составляет {:.0%} от общего количества данных'.format(data_percent_na))
    if data_percent_na > 0.1:
        print('Вывод: число пропущенных значений составляет более 1%, следовательно, удалить строки с пропущенными значениями нельзя, их следует заполнить и использовать в анализе')
    elif data_percent_na == 0:
        print('Вывод: в таблице данных нет пропущенных значений, либо пропуски уже были заполнены')
    else:
        print('Вывод: число пропущенных значений составляет менее 1%, следовательно, этими данными можно пренебречь и удалить их')
except:
    print('Ошибка при расчёте, убедитесь, что данные были импортированы корректно')
    

# Заполнение пропущенных значений
# Поскольку пропущенные значения столбца days_employed в решении поставленных в проекте задач ценности не имеют, допустимо заполнить их нулями
data['days_employed'] = data['days_employed'].fillna(0)    # Применение метода fillna с аргументом 0 к столбцу days_employed

# Поскольку нельзя определить ежемесячный доход для пропущенных значений, их следует заполнить условным значением, например, -1,
# и в дальнейшем анализе по столбцу total_income их не учитывать, а в анализе по другим столбцам - учитывать.
data['total_income'] = data['total_income'].fillna(-1)    #  Применение метода fillna с аргументом -1 к столбцу total_income

# Проверка результатов заполнения пропусков
print()
print('Проверка результатов заполнения пропусков:')
data.info()


Общее количество строк в исходной таблице составляет: 21525
Общее количество пропущенных значений в столбце "ежемесячный доход" составляет: 2174
Число пропущенных значений в исходной таблице составляет 10% от общего количества данных
Вывод: число пропущенных значений составляет более 1%, следовательно, удалить строки с пропущенными значениями нельзя, их следует заполнить и использовать в анализе

Проверка результатов заполнения пропусков:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-nu

### Вывод

1. На первом шаге было определено, что пропуски значений есть в столбцах **days_employed** и **total_income**.
1. Пропуски были обнаружены методом **isna()**. Обработаны методом **fillna()**.
1. Пропущенные значения являются NaN, значит с ними возможны математические операции.
1. Пропуски значений в столбцах **days_employed** и **total_income** наблюдаются в одних и тех же строках, другими словами, для части клиентов данные по двум столбцам не были заполнены, либо потерялись в процессе выгрузки данных.
1. Пропуски для столбца **days_employed** были заполнены нулями, т.к. значения данного столбца для решения задач проекта не интересны, но было необходимо сохранить значения остальных столбцов.
1. Пропуски для столбца **total_income** были заполнены значением -1, чтобы в дальнейшем можно было исключить эти значения из выборки по условию.

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>
    
Такая обработка маркерныеми значениями подойдет, т.к. мы не используем далее их в аналихе. Однако лучше использовать заполнение пропусков в соответствии с типом дохода. Тогда пенсионеры будут получать также, как и другие пенсионеры и т.д.

### Замена типа данных
- Заменить вещественный тип данных на целочисленный: пояснить, как выбирался метод для изменения типа данных;

***Примечание.*** В данных могут встречаться артефакты — значения, которые не отражают действительность. Например, отрицательное количество дней трудового стажа. Для реальных данных — это нормально. Нужно описать возможные причины появления таких данных и обработать их.

In [5]:
# Преобразование типов данных
data['days_employed'] = data['days_employed'].astype('int')    # Преобразование типа данных из вещественного в целочисленный методом astype с аргументов 'int' для столбца days_employed
data['total_income'] = data['total_income'].astype('int')    # Преобразование типа данных из вещественного в целочисленный методом astype с аргументов 'int' для столбца total_income

# Проверка выполнения преобразования типов данных
print(data['days_employed'].head())    # Вывести на экран первые пять строк по столбцу days_employed
print(data['total_income'].tail())    # Вывести на экран последние пять строк по столбцу total_income

0     -8437
1     -4024
2     -5623
3     -4124
4    340266
Name: days_employed, dtype: int64
21520    224791
21521    155999
21522     89672
21523    244093
21524     82047
Name: total_income, dtype: int64


### Вывод

1. На первом шаге было определено, что вещественный тип данных присутствует в столбцах **days_employed** и **total_income**.
1. Для преобразования типа данных был использован метод **astype()** с аргументом **'int'**, так как функция **to_numeric()** преобразовывает из ***строкового*** типа в ***вещественный***, а нам нужно из ***вещественного*** в ***целочисленный***.
1. В столбце **days_employed** встречаются отрицательные значения которых быть не должно. Вероятно это системная ошибка которая возникла при выгрузке данных на стороне заказчика.  Поскольку данный столбец для решения поставленных проектом задач ценности не имеет, то следует проигнорировать отрицательные значения, т.е. ничего с ними не делать. В отчёте следует указать заказчику на данную проблему, чтобы они нашли её причину и устранили.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена верно. 

### Обработка дубликатов

1. Удалить дубликаты: пояснить, как выбирался метод для поиска и удаления дубликатов в данных;
1. Привести возможные причины появления дубликатов;

In [6]:
# На первом шаге было выяснено, что в столбце education не соблюдается регистр.
# Необходимо привести значения в столбце education к нижнему регистру и затем проверить наличие дубликатов
data['education'] = data['education'].str.lower()    # Применение метода Pandas str.lower() для приведения строк в колонке education к нижнему регистру

# Проверим результат применения метода приведения строк к нижнему регистру
print()    # Отступ 
print('После применения метода приведения строк к нижнему регистру в столбце "образование" получили следующий набор значений:')
print(data['education'].unique())    # Вывод на экран всех уникальных значений столбца education

# Далее следует проверить данные на наличие полных дубликатов
data_duplicates_sum = data.duplicated().sum()    # Применение метода определения дубликатов duplicated() и метода подсчёта их количества sum() и запись результата в переменную data_duplicates_sum
print('Количество полных дубликатов в исходных данных составляет:', data_duplicates_sum)    # Вывод результата операции на экран

# Затем следует удалить найденные полные дубликаты
data = data.drop_duplicates().reset_index(drop = True)    # Применение метода drop_duplicates() с одновременным сбросом индексации строк и записью результата в исходную таблицу data
print('Общее количество строк в исходной таблице составляло:', data_rows_total)    # Вывод результата операции на экран
print('После удаления полных дубликатов количество строк уменьшилось на', data_duplicates_sum,'и теперь составляет:', data.shape[0])    # Вывод результата операции на экран

# Убедимся, что не осталось дубликатов в полученных данных
print('Количество полных дубликатов в полученных данных составляет:', data.duplicated().sum())   # Вывод результата операции на экран



После применения метода приведения строк к нижнему регистру в столбце "образование" получили следующий набор значений:
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']
Количество полных дубликатов в исходных данных составляет: 71
Общее количество строк в исходной таблице составляло: 21525
После удаления полных дубликатов количество строк уменьшилось на 71 и теперь составляет: 21454
Количество полных дубликатов в полученных данных составляет: 0


### Вывод

1. В данных был обнаружен и удалён 71 дубликат.
1. Для поиска и подсчёта количества дубликатов были использованы методы **duplicated()** вместе с **sum()**. Метод **duplicated()** выбирался как единственно возможный применительно к данныму проекту, поскольку в исходной таблице нет ID, ФИО, названий или других подобных категорийных столбцов по которым можно было бы провести сравнение и применить метод **unique()**.
1. Для удаления дубликатов был использован метод **drop_duplicates()**. По той же причине.
1. Дубликаты могли появиться в исходных данных из-за неправильного объединения данных из разных источников (криво скопипастили несколько табличек в одну), ошибки оператора при внесении информации (записали в базу одни и те же данные два раза подряд потому что первый раз показалось что "виндоуз завис" плюс юзабилити интерфейса на уровне Windows 98).

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.

### Лемматизация

1. Выделить леммы в значениях столбца с целями получения кредита: описать, как проводилась лемматизация целей кредита;

In [7]:
from collections import Counter    # Импорт коллекции Counter
from pymystem3 import Mystem    # Импорт библиотеки Mystem
m = Mystem()    # переменной m присвоить Mystem() 


# Выделим список уникальных значений строк в столбце purpose и выполним с ним лемматизацию
data_purpose_unique = data['purpose'].unique()    # Выявление уникальных строк в столбце purpose с последующей записи результата в переменную data_purpose_unique
lemmas = m.lemmatize(' '.join(data_purpose_unique))    # Выполнение леммматизации функцией lemmatize() с аргументом в виде строки, полученной методом join() с аргументом в виде переменной data_purpose_unique
print(Counter(lemmas))    # Подсчёт количества повторяющихся лемм с помошью функции Counter() и вывод результата на экран
print()

# Исходя из обнаруженных наиболее часто встречающихся слов в столбце purpose можно определить категории целей получения кредита
def purpose_lemmatize(row):    # Функция которая возвращает наименование категории в зависимости от найденных в исходном столбце слов
    lemmas = m.lemmatize(row)    # Привести к лемме строковое значение столбца
    if 'недвижимость' in lemmas or 'жилье' in lemmas or 'строительство' in lemmas:   # Если встречаются перечисленные слова, то вернуть название соответствующей им категории
        return 'недвижимость'
    elif 'автомобиль' in lemmas:
        return 'автомобиль'
    elif 'образование' in lemmas:
        return 'образование'
    else:    # Наименее часто встречающиеся цели кредита (свадьба, ремонт) попадают в категорию "прочее"
        return 'прочее'

# Применим написанную функцию к столбцу purpose. Выполнение функции занимает примерно 3-4 секунды
data['purpose_category'] = data['purpose'].apply(purpose_lemmatize)     # С помощью метода apply() применяем функцию
#purpose_lemmatize к столбцу purpose, а результат записываем в новый столбец purpose_category
 
# Выполним подсчёт количества клиентов по категориям целей кредита
print(Counter(data['purpose_category']))



# Пробовал стемминг. Результат не устроил, поэтому оставил в комментариях.
#Стемминг дал неточный результат и операция занимает много больше времени по сравнению с лемматизацией и методом apply
#from nltk.stem import SnowballStemmer 
#russian_stemmer = SnowballStemmer('russian')
#for i in range(data.shape[0]):
#    for word in data.loc[i, 'purpose'].split():
#        stemmed_word = russian_stemmer.stem(word)
#        if stemmed_word == 'недвижим' or stemmed_word == 'жил' or stemmed_word == 'строительств':    # жил    автомобил   строительств
#            data.loc[i, 'purpose'] = 'недвижимость'
#        elif stemmed_word == 'автомобил':
#            data.loc[i, 'purpose'] = 'автомобиль'
#        elif stemmed_word == 'образован':
#            data.loc[i, 'purpose'] = 'образование'
#        else:
#            data.loc[i, 'purpose'] = 'прочее'
#Out:     'недвижимость': 9522, 'прочее': 4091, 'образование': 4013, 'автомобиль': 3828})


Counter({' ': 96, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'подержать': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'со': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1, '\n': 1})

Counter({'недвижимость': 10811, 'автомобиль': 4306, 'образование': 4013, 'прочее': 2324})


### Вывод

1. На первом шаге было определено, что в столбце **purpose** встречаются 38 уникальных строковых значений, но по смыслу большинство из них соответствуют одной и той же цели.
1. Для уменьшения количества целей до нескольких основных категорий была применена лемматизация.
1. В результате из 38 целей получения кредита выделили 4 категории: недвижимость - встречается 10811 раз, автомобиль - 4306 раз, образование - 4013 раз, прочее - 2324 раза.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Все верно, метод lemmatize() использован правильно. Выбор лемм объяснен. Только лемма прочее состоит из одной глобальной цели, уточни ее. 

### Категоризация данных

- Категоризировать данные: перечислить, какие «словари» выделили для этого набора данных, и объяснить, почему.

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Категоризовать данные по числу детей необязательно. Разбиение есть уже в исходных данных. Все-таки разница в возврате кредита у семей с одним ребенком или тремя детьми будет.


In [8]:
# В начале описаны функции которые будут задействованы при категоризации
def children_cat(row):  # Функция для категоризации клиентов по количеству детей
    children = row
    if children == 0:     # Если количество детей равно нулю, значит нет детей
        return 'нет детей'
    elif 1 <= children <= 3:   # Если детей от 1 до 3 включительно будем считать что дети есть
        return 'есть дети'
    elif 4 <= children <= 10:    # Если детей от 4  до 10 включительно будем считать многодетной семьей
        return 'многодетная семья'
    return 'неверные данные'    # Во всех остальных случаях будем считать что данные указаны неверно


def total_income_cat(row):    # Функция для категоризации клиентов по ежемесячному доходу
    total_income = row
    if 0 < total_income <= 100000:    # Если ежемесячный доход в диапазоне от 0 до 100 тыс. то попадает с эту категорию
        return 'От 0 до 100 тыс.'
    elif 100000 < total_income <= 140000:
        return 'От 100 до 140 тыс.'
    elif 140000 < total_income <= 190000:
        return 'От 140 до 190 тыс.'
    elif 190000 < total_income <= 330000:
        return 'От 190 до 330 тыс.'
    elif total_income > 330000:
        return 'От 330 тыс. и выше'
    return 'Пропущенные данные'

    
# 1. Категоризация по столбцу "количество детей"
# Исследуем какое количество детей встречается в данных
print('В столбце "количество детей" встречаются следующие  значения:')
print(Counter(data['children']))    # Применение функции Counter для подсчета количества различных значений столбца children
print('Обнаружены выбросы: 20: 76, -1: 47, что составляет соответственно 0,35% и ')
print('0,22% от количества строк в таблице. Выбросы будут помещены в категорию "неверные данные"')
print()

# Категоризируем данные по столбцу "количество детей" и запишем результат в новый столбец 
data['children_category'] = data['children'].apply(children_cat)    # Применение метода apply() к столбцу children с функцией children_cat

# Выведем суммарные значения по новому столбцу children_category на экран для оценки
print('Результат работы функции категоризации по количеству детей:')
print(data['children_category'].value_counts())
print()

# 2. Категоризация по столбцу "ежемесячный доход"
# Исследуем какие суммы доходов и как часто встречаются в данных с помощью выборки из DataFrame с условием и методом count()
print('В столбце "ежемесячный доход" можно выделить следующие категории сумм которые встречаются с примерно одинаковой частотой:')
print('От 0 до 100 тыс.:', data[(data['total_income'] > 0) & (data['total_income'] <=100000)]['total_income'].count())
print('От 100 до 140 тыс.:', data[(data['total_income'] > 100000) & (data['total_income'] <=140000)]['total_income'].count())
print('От 140 до 190 тыс.:', data[(data['total_income'] > 140000) & (data['total_income'] <=190000)]['total_income'].count())
print('От 190 до 330 тыс.:', data[(data['total_income'] > 190000) & (data['total_income'] <=330000)]['total_income'].count())
print('От 330 тыс. и выше:', data[data['total_income'] > 330000]['total_income'].count())
print('Пропущенные данные:', data[data['total_income'] == -1]['total_income'].count())
print()

# Категоризируем данные по столбцу "ежемесячный доход" и запишем результат в новый столбец 
data['total_income_category'] = data['total_income'].apply(total_income_cat)    # Применение метода apply() к столбцу total_income с функцией total_income

# Выведем суммарные значения по новому столбцу total_income_category на экран для оценки
print('Результат работы функции категоризации по ежемесячному доходу:')
print(data['total_income_category'].value_counts())


В столбце "количество детей" встречаются следующие  значения:
Counter({0: 14091, 1: 4808, 2: 2052, 3: 330, 20: 76, -1: 47, 4: 41, 5: 9})
Обнаружены выбросы: 20: 76, -1: 47, что составляет соответственно 0,35% и 
0,22% от количества строк в таблице. Выбросы будут помещены в категорию "неверные данные"

Результат работы функции категоризации по количеству детей:
нет детей            14091
есть дети             7190
неверные данные        123
многодетная семья       50
Name: children_category, dtype: int64

В столбце "ежемесячный доход" можно выделить следующие категории сумм которые встречаются с примерно одинаковой частотой:
От 0 до 100 тыс.: 4463
От 100 до 140 тыс.: 4642
От 140 до 190 тыс.: 4496
От 190 до 330 тыс.: 4649
От 330 тыс. и выше: 1101
Пропущенные данные: 2103

Результат работы функции категоризации по ежемесячному доходу:
От 190 до 330 тыс.    4649
От 100 до 140 тыс.    4642
От 140 до 190 тыс.    4496
От 0 до 100 тыс.      4463
Пропущенные данные    2103
От 330 тыс. и выше   

### Вывод

1. В соответствии с поставленными в данном проекте вопросами категоризация была выполнена для следующих столбцов: **children**,  **total_income** и **purpose** - последний был категоризирован на шаге "Лемматизация", см. выше.
1. На первом шаге приняли, что столбцу **children** следует применить категоризацию следующего вида: 0 - нет детей, от 1 до 3 - есть дети, от 4 и более - многодетная семья. Поскольку в данных столбца обнаружились выбросы, то последнее условие было ограничено от 4 до 10 детей - многодетная семья. Таким образом, все данные за пределами данной выборки были обозначены как "неверные данные" и будут учитываться в отдельной категории.
1. По столбцу **total_income** были выделены 6 категорий по сумме дохода:  От 0 до 100 тыс., от 100 до 140 тыс., от 140 до 190 тыс., от 190 до 330 тыс., от 330 тыс. и выше, и пропущенные данные. Категории были определены так, чтобы в каждую попадало примерно одинаковая выборка данных.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Для того чтобы в каждую выделяемую группу попало одинаковое число клиентов, можно использовать метод qcut, который разделит данные по квантилям на указанное число групп. В остальном все сделано правильно. 

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [9]:
# Функция для подготовки сводной таблицы для отчёта, в качестве аргумента задается название столбца с категориями
def data_pivot_fun(column_name):    
    # Обработаем данные с помощью метода pivot
    # Создание сводной таблицы из столбцов <column_name>, debt, с подсчётом количества значений функцией count 
    data_pivot = data.pivot_table(index=column_name, columns='debt', values='family_status_id', aggfunc='count')

    # Изменение названий столбцов для удобства восприятия итоговой таблицы методом set_axis
    data_pivot.set_axis(['нет долгов','есть долги'],axis = 'columns',inplace = True)

    # Добавление в сводную таблицу столбца с подсчётом общего количества клиентов
    data_pivot['всего клиентов'] = data_pivot['нет долгов'] + data_pivot['есть долги']

    # Добавление в сводную таблицу столбца с расчётом отношения должников к общему количеству клиентов с применением форматирования методом apply()
    try:
        data_pivot['% должников'] = (data_pivot['есть долги'] / data_pivot['всего клиентов']).apply('{:.2%}'.format) 
    except:
        return 'Не удалось вычислить отношение, попробуйте запустить программу заново'

    # Вывод сводной таблицы на экран с сортировкой по столбцу "% должников" по-убыванию
    return data_pivot.sort_values(by = '% должников', ascending = False)

# Вызов функции data_pivot_fun с аргументом children_category
display(data_pivot_fun('children_category'))

нет долгов  есть долги  всего клиентов % должников
children_category                                                    
есть дети                6525         665            7190       9.25%
многодетная семья          46           4              50       8.00%
нет детей               13028        1063           14091       7.54%
неверные данные           114           9             123       7.32%

### Вывод

1. На основании приведённых результатов можно сказать, что наличие у клиента детей не сильно, но всё же негативно влияет на платёжеспособность: среди клиентов с детьми на 1,71% должников больше, чем среди клиентов без детей.
1. Касательно клиентов с многодетными семьями - их общее количество в представленных данных слишком мало: 0,23% от всех данных, это даже меньше чем количество выбросов. Следовательно, по ним вывод сделать нельзя, а рассчетное значение получилось как раз в интервале между двумя другими категориями.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Вывод не противоречит полученному результату. 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [10]:
# Вызов функции data_pivot_fun с аргументом family_status
display(data_pivot_fun('family_status'))



нет долгов  есть долги  всего клиентов % должников
family_status                                                            
Не женат / не замужем        2536         274            2810       9.75%
гражданский брак             3763         388            4151       9.35%
женат / замужем             11408         931           12339       7.55%
в разводе                    1110          85            1195       7.11%
вдовец / вдова                896          63             959       6.57%

### Вывод

1. По результатам анализа можно сказать, что зависимость между семейным положением и возвратом кредита в срок есть.
1. Должников меньше всего среди вдовцов и вдов
1. Процент должников среди неженатых/незамужних и тех кто в гражданском браке примерно одинаковый и он на 2,2% больше чем у женатых/замужних и разведенных.
1. Наибольший риск для банка представляет категория неженатых/незамужних клиентов.

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Таблицы на печать выводи с помощью метода display. 

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [11]:
# Вызов функции data_pivot_fun с аргументом total_income_category
display(data_pivot_fun('total_income_category'))

нет долгов  есть долги  всего клиентов % должников
total_income_category                                                    
От 140 до 190 тыс.           4089         407            4496       9.05%
От 100 до 140 тыс.           4250         392            4642       8.44%
Пропущенные данные           1933         170            2103       8.08%
От 0 до 100 тыс.             4109         354            4463       7.93%
От 190 до 330 тыс.           4309         340            4649       7.31%
От 330 тыс. и выше           1023          78            1101       7.08%

### Вывод

1. Зависимость между уровнем дохода и сроком возврата кредита в срок получилась нелинейная: в начале по мере роста уровня дохода от 0 до 190 тыс. процент должников растёт примерно от 8% до 9%, а с дальнейшим ростом уровня дохода от 190 тыс. и выше процент должников уменьшается с 9% до 7%.
1. Наибольший риск для банка представляет категория с с уровнем дохода от 140 до 190 тыс.

- Как разные цели кредита влияют на его возврат в срок?

In [12]:
# Вызов функции data_pivot_fun с аргументом purpose_category
display(data_pivot_fun('purpose_category'))

нет долгов  есть долги  всего клиентов % должников
purpose_category                                                    
автомобиль              3903         403            4306       9.36%
образование             3643         370            4013       9.22%
прочее                  2138         186            2324       8.00%
недвижимость           10029         782           10811       7.23%

### Вывод

1. Если цели кредита связаны с покупкой автомобиля или образованием, то процент должников в этих категориях примерно одинаковый и самый большой из представленных.
1. Меньше всего должников в категории "недвижимость" - примерно на 2% меньше по сравнению с автомобилями и образованием. Причём количество клиентов в данной категории наибольшее (примерно равно сумме клиентов во всех остальных категориях).
1. Процент должников в категории "прочее" оказался посередине между недвижимостью и автомобилями с образованием.

<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

Здорово, что при анализе использован метод сводных таблиц. Он пригодится в дальнейших проектах.

Выводы верные по всем пунктам. 

### Шаг 4. Общий вывод

- По результатам проведенного анализа можно заключить, что семейное положение и количество детей у клиента оказывают влияние на погашение кредита в срок: наибольший риск представляют неженатые/незамужние клиенты с детьми, наименьший - состояющие в браке и без детей.

<div style="border:solid  orange  2px; padding: 20px"> <h1 style="color: orange">Комментарий наставника</h1>

Выводы описаны кратко. Однако приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения. Также можно расписать все, что было сделано в ходе проведения работы. 


<div style="border:solid  green  2px; padding: 20px"> <h1 style="color: green">Комментарий наставника</h1>

- Таблицы на печать выводи с помощью метода display;
- Финальный вывод стоит писать подробно.

Ты проделал большую работу, молодец! Критических замечаний нет. Работа выполнена хорошо. Продолжай в том же духе :)

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.